test/ipcAPI.cpp

```diff
@@ -0,0 +1,336 @@
+        // TODO: The function is similar to poolCreateExt function
+        //       from memoryPool.hpp
+        umf_memory_provider_handle_t hProvider;
+        umf_memory_pool_handle_t hPool;
+
+        auto ret =
+            umfMemoryProviderCreate(&IPC_MOCK_PROVIDER_OPS, &stat, &hProvider);
+        EXPECT_EQ(ret, UMF_RESULT_SUCCESS);
+
+        ret = umfPoolCreate(&umf_test::PROXY_POOL_OPS, hProvider, nullptr,
+                            &hPool);
+        EXPECT_EQ(ret, UMF_RESULT_SUCCESS);
+
+        // capture provider and destroy it after the pool is destroyed
+        auto poolDestructor = [hProvider](umf_memory_pool_handle_t pool) {
+            umfPoolDestroy(pool);
+            umfMemoryProviderDestroy(hProvider);
+        };
+
+        return umf::pool_unique_handle_t(hPool, std::move(poolDestructor));
+    }
+
+    using stats_type = typename provider_mock_ipc::stats;
+    umf_memory_provider_ops_t IPC_MOCK_PROVIDER_OPS =
+        umf::providerMakeCOps<provider_mock_ipc, stats_type>();
+    umf::pool_unique_handle_t pool;
+    static constexpr int NTHREADS = 10;
+    stats_type stat;
+};
+
+TEST_F(umfIpcTest, BasicFlow) {
+    constexpr size_t SIZE = 100;
+    int *ptr = (int *)umfPoolMalloc(pool.get(), SIZE * sizeof(int));
+    EXPECT_NE(ptr, nullptr);
+
+    std::iota(ptr, ptr + SIZE, 0);
+
+    umf_ipc_handle_t ipcHandleFull = nullptr;
+    size_t handleFullSize = 0;
+    umf_result_t ret = umfGetIPCHandle(ptr, &ipcHandleFull, &handleFullSize);
+    ASSERT_EQ(ret, UMF_RESULT_SUCCESS);
+    umf_ipc_handle_t ipcHandleHalf = nullptr;
+    size_t handleHalfSize = 0;
+    ret = umfGetIPCHandle(ptr + SIZE / 2, &ipcHandleHalf, &handleHalfSize);
+    ASSERT_EQ(ret, UMF_RESULT_SUCCESS);
+    ASSERT_EQ(handleFullSize, handleHalfSize);
+
+    void *fullArray = nullptr;
+    ret = umfOpenIPCHandle(pool.get(), ipcHandleFull, &fullArray);
+    ASSERT_EQ(ret, UMF_RESULT_SUCCESS);
+
+    void *halfArray = nullptr;
+    ret = umfOpenIPCHandle(pool.get(), ipcHandleHalf, &halfArray);
+    ASSERT_EQ(ret, UMF_RESULT_SUCCESS);
+
+    for (int i = 0; i < (int)SIZE; ++i) {
+        ASSERT_EQ(reinterpret_cast<int *>(fullArray)[i], i);
+    }
+    // Close fullArray before reading halfArray
+    ret = umfCloseIPCHandle(fullArray);
+    EXPECT_EQ(ret, UMF_RESULT_SUCCESS);
+
+    for (int i = 0; i < (int)SIZE / 2; ++i) {
+        ASSERT_EQ(reinterpret_cast<int *>(halfArray)[i], i + SIZE / 2);
+    }
+    ret = umfCloseIPCHandle(halfArray);
+    EXPECT_EQ(ret, UMF_RESULT_SUCCESS);
+
+    ret = umfPutIPCHandle(ipcHandleFull);
+    EXPECT_EQ(ret, UMF_RESULT_SUCCESS);
+
+    ret = umfPutIPCHandle(ipcHandleHalf);
+    EXPECT_EQ(ret, UMF_RESULT_SUCCESS);
+
+    ret = umfPoolFree(pool.get(), ptr);
+    EXPECT_EQ(ret, UMF_RESULT_SUCCESS);
+
+    EXPECT_EQ(stat.getCount, 1);
+    EXPECT_EQ(stat.putCount, stat.getCount);
+    // EXPECT_EQ(stat.openCount, 1);
+    EXPECT_EQ(stat.closeCount, stat.openCount);
+}
+
+TEST_F(umfIpcTest, ConcurrentGetPutHandles) {
```

Comment:
nit: it would be good to have a multithreaded tests with a provider that does not use any locks internally. With provider_mock_ipc you introduced extra synchronization which can hide some potential issues with critnib I think.

Reply:
Agree. I will think about that.